text/0000-module-resolution.md

```diff
@@ -0,0 +1,458 @@
+               "common-util:": "bar-common-utils"
+           }
+       }     
+    ]
+}
+```
+
+::: warning
+The module resolver will throw if it detects multiple modules with the same name
+:::
+
+
+## Algorithms
+
+### Preload all modules
+
+Here is the algoritm for module resolution (the instructions are not in a rigurous specification format for brevity and time)
+
+```
+  Let `rootDir` be the initial path for the application, `initModules` 
+  the list of ModuleRecords to be resolved and ModuleRecordEntryList 
+  an empty array.
+  
+  1. Load the LWC configuration file in the rootDir
+    1.1 Search for `lwc.config.js`, if it exist read it.
+    1.2 If `lwc.config.js` is not found search in `package.json`
+    (If both are found the `lwc.config.js` file takes precedence).
+
+  2. Merge all the ModuleRecords found in the config with 
+  the list of initModules. If an alias with the same name 
+  exist pick initModules over the ones resolved in the config.
+
+  3. For each ModuleRecord:
+
+    3.1  Validate that its a valid ModuleRecord type, skip it otherwise.
+    3.1.1 For `AliasModuleRecord` `path` and `name` keys must exist.
+    3.1.2 For `DirModuleRecord` the dir key must exist and the 
+    value must match an existing folder on the file system.
+    3.1.3 For `NpmModuleRecord` the npm key must exist and its value 
+    must not have a leading `/` or `./`. Also the npm package must
+    be resolvable from Node's perspective.
+
+    3.2 Resolve ModuleRecord entry:
+
+        3.2.1 If is an `AliasModuleRecord` validate the path and add
+        create a ModuleRecordEntry with:
+          - `specifier` as the `name` value.
+          - `scope` with the closest configuration path.
+          - `entry` with the `path` value.
+
+        3.2.2 Else if is a `DirModuleRecord` validate the path and find of 
+        the modules that match the structure:
+        [namespace]/[componentName]/[componentName.{html|css|js|ts}] and create a 
+        ModuleRecordEntry with:
+          - `specifier` be `[namespace]/[componentName]`
+          - `scope` be the closest configuration path
+          - `entry` be `[dir]/[namespace]/[componentName]/[componentName.{html|css|js|ts}]`
+
+        3.2.3 Else if is a `NpmModuleRecord`, validate that the npm package
+        is installed and can be resolved (by using Node `node_modules` 
+        algorithm). Then `GoTo 1` recursively merging all the 
+        ModuleRecordEntries collecting the scope forEach as npm 
+        packages are traversed.
+3.2.4 Otheriwse throw an invalid ModuleRecord error
+    3.3 Return all the ModuleRecordEntries collected.
+```
+
+### Algorithm to resolve a module in iterable way
+
+This algorithm is very similar to the [NodeJS require](https://nodejs.org/api/modules.html#modules_all_together),
+the main difference is that we don't traverse upwards all module paths. 
+
+```
+Lets `importee` be module specifier to be resolved and `importer` the path of the module on 
```

Comment:
We should add a step here to check that the `importee` is a bare import (identifier with no `/` or `./` prefix), otherwise, the algorithm should early exit.